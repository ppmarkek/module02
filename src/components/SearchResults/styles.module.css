.searchResults {
  height: 100%;
  width: 50%;
  border-radius: 15px;
  background: #f8f9fc;
  padding: 5px;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  overflow: auto;
  border: grey 1px solid;
}

.searchResultsWrapper {
  height: 85%;
  width: 100%;
  padding-bottom: 10px;
  display: flex;
  gap: 10px;
  flex-direction: column;
  align-items: center;
}

.error {
  display: block;
}
