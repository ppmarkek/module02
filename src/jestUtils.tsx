import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { reducer } from './redux/index';
import { createStore } from '@reduxjs/toolkit';

export const renderWithRedux = (ui: React.JSX.Element, state = {}) => {
  const Wrapper = ({
    children,
  }: {
    children: React.ReactNode | React.ReactNode[];
  }) => (
    <Router>
      <Provider store={createStore(reducer, state)}>{children}</Provider>
    </Router>
  );

  return render(ui, { wrapper: Wrapper });
};
