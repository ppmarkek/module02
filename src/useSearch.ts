import { useSelector } from 'react-redux';
import { AppState } from './types';

export const useSearch = () => {
  const requestResults = useSelector(
    (state: AppState) => state.componentsReducer.results
  );
  const isLoading = useSelector(
    (state: AppState) => state.componentsReducer.isLoading
  );
  const details = useSelector(
    (state: AppState) => state.componentsReducer.details
  );
  const searchResult = useSelector(
    (state: AppState) => state.componentsReducer.searchResult
  );

  const isDetailsLoading = useSelector(
    (state: AppState) => state.componentsReducer.isDetailsLoading
  );

  const isShowDetails = useSelector(
    (state: AppState) => state.componentsReducer.isShowDetails
  );

  const isError = useSelector(
    (state: AppState) => state.componentsReducer.isError
  );

  return {
    requestResults,
    isLoading,
    isDetailsLoading,
    details,
    searchResult,
    isShowDetails,
    isError,
  };
};
