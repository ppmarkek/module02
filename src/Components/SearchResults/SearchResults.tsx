import { SearchItemCard } from '../SearchItemCard/SearchItemCard';
import { Result } from '../../types';
import './styles.css';
import { useSearch } from '../../useSearch';

export const SearchResults = () => {
  const { requestResults, isLoading, isError } = useSearch();
  const emptyResult = Object.keys(requestResults).map((value, index) => (
    <p key={index} data-testid={`empty-results-categories-${index}`}>
      <h4>{value}</h4>
    </p>
  ));

  return (
    <div className="searchResultsWrapper">
      <div
        className={`searchResults ${isError && 'error'}`}
        data-testid="search-results"
      >
        {isLoading && <h3 data-testid="loading">Loading...</h3>}
        {isError && !isLoading && (
          <h2 data-testid="error">
            Nothing found by selected criteria, try this keywords:{' '}
          </h2>
        )}
        {isError && !isLoading && emptyResult}
        {!isError &&
          !isLoading &&
          requestResults?.results?.map((result: Result, index: number) => (
            <SearchItemCard key={index} details={result} number={index} />
          ))}
      </div>
    </div>
  );
};
