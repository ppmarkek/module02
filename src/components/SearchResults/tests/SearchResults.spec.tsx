import { SearchResults } from '../SearchResults';
import { renderWithRedux } from '../../../jestUtils';
import { appStoreMock, useSearchHookMock } from '../../../fixtures/results';
import * as useSearch from '../../../useSearch';
import '@testing-library/jest-dom';

describe('SearchResult component', () => {
  it('should render correctly', () => {
    jest
      .spyOn(useSearch, 'useSearch')
      .mockImplementation(() => useSearchHookMock);

    const component = renderWithRedux(<SearchResults />, {
      appStore: appStoreMock,
    });

    expect(component).toMatchSnapshot();
  });

  it('should display cards', () => {
    const { getByTestId } = renderWithRedux(<SearchResults />, {
      state: appStoreMock,
    });

    expect(getByTestId('search-item-card-1'));
  });

  it('should display loading', () => {
    jest.spyOn(useSearch, 'useSearch').mockImplementationOnce(() => ({
      ...useSearchHookMock,
      isLoading: true,
    }));

    const { getByTestId } = renderWithRedux(<SearchResults />);

    expect(getByTestId('loading'));
  });

  it('should display categories', () => {
    jest.spyOn(useSearch, 'useSearch').mockImplementationOnce(() => ({
      ...useSearchHookMock,
      isLoading: false,
      isError: true,
    }));

    const { getByTestId } = renderWithRedux(<SearchResults />);

    expect(getByTestId('empty-results-categories-1'));
  });
});
