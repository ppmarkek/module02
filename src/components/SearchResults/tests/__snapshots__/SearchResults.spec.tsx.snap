// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="searchResultsWrapper"
      >
        <div
          class="searchResults false"
          data-testid="search-results"
        >
          <div
            class="searchItemCard"
            data-testid="search-item-card-0"
          >
            <h3>
              Name: 
              test name
            </h3>
            <p>
              Created: 
              01-01-0101
            </p>
            <p>
              Edited: 
              02-02-0202
            </p>
            <p>
              Height: 
              100
            </p>
            <p>
              Mass: 
              100
            </p>
          </div>
          <div
            class="searchItemCard"
            data-testid="search-item-card-1"
          >
            <h3>
              Name: 
              test name
            </h3>
            <p>
              Created: 
              02-02-0202
            </p>
            <p>
              Edited: 
              03-03-0303
            </p>
            <p>
              Height: 
              test height
            </p>
            <p>
              Mass: 
              test mass
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="searchResultsWrapper"
    >
      <div
        class="searchResults false"
        data-testid="search-results"
      >
        <div
          class="searchItemCard"
          data-testid="search-item-card-0"
        >
          <h3>
            Name: 
            test name
          </h3>
          <p>
            Created: 
            01-01-0101
          </p>
          <p>
            Edited: 
            02-02-0202
          </p>
          <p>
            Height: 
            100
          </p>
          <p>
            Mass: 
            100
          </p>
        </div>
        <div
          class="searchItemCard"
          data-testid="search-item-card-1"
        >
          <h3>
            Name: 
            test name
          </h3>
          <p>
            Created: 
            02-02-0202
          </p>
          <p>
            Edited: 
            03-03-0303
          </p>
          <p>
            Height: 
            test height
          </p>
          <p>
            Mass: 
            test mass
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
